### ISSUE TITLE
Error report with the filter Morph (not using the good Ubuntu version...)

### ISSUE THREAD

**Issue by fabiengabriel** [GreycLab/gmic#15]
Hi
Just to report an error message with the filter Morph [Interactive] :
```
*** Error in ./fx_morph_interactive/x_morph/*repeat/*local/*substitute/max_wh/_minmax_whds/ *** Item substitution '{ mw = w; mh = h; md = d; ms = (...)s#k); ); ([mw,mh,md,ms])[0,2]}': Unrecognized function call 'repeat( l,k, mw = max(mw,w#k); mh = max(mh,h#k); md = max(m(...)' in expression '...repeat( l,k, mw = max(mw,w#k); mh = max(mh,h#k); md = max(m(...)...'.
```
I use the last version 3.3.0 of "Ubuntu 22.04 Jammy: Plug-in for GIMP 2.10" with GIMP 2.10.30 on ... Ubuntu 23.04 ... maybe is it the cause of the issue? (But many other filters work anyway)
If it is just because of the Ubuntu version, sorry! But if not, I hope my report could be helpful.
Thanks!

**Comment by dtschump** [comment]
Hello.
If the plug-in runs, then using it in 23.04 is not the cause of the issue.
I've tried here, and this filter works as expected. Have you tried refreshing your filters ?
If so, could you provide a simple and detailed way to reproduce the issue (image data and steps ?).
Thanks!

**Comment by fabiengabriel** [comment]
Hi

The error remain after refreshing the filters.

I tried different images. Here the two pictures I used for the test :
- https://unsplash.com/fr/photos/NPDnHmYsl6Y
- https://unsplash.com/fr/photos/s6g6ZSxM3kQ

I go to the menu "Filters">G'MIC-Qt..." which opens a window. I refresh the filters (the icon of a spinning arrow, with the box "Internet" checked), and it says : "Les définitions de filtres ont été mises à jour".

![Capture d’écran du 2023-09-21 13-58-35](https://github.com/GreycLab/gmic/assets/41618686/027b883f-7f8f-4910-90c4-0fe9ba894039)

With some basic transformations, the filters are okay:

![Capture d’écran du 2023-09-21 13-54-21](https://github.com/GreycLab/gmic/assets/41618686/af7ae7c2-0db3-4549-8213-f2e07f47be2d)

Here a screenshot of the error with "Morph":

![Capture d’écran du 2023-09-21 13-54-59](https://github.com/GreycLab/gmic/assets/41618686/b40c38ca-0c16-44f9-8680-d756d4b3c5ef)

I see that the error appears not only with the "Morph" filter, but also with many others... Even "What's New?":

![Capture d’écran du 2023-09-21 14-00-32](https://github.com/GreycLab/gmic/assets/41618686/630d1272-f4a6-4bef-a9ff-7bb28e2f8cfb)

I test the error with the functions : 
- Array [Random] : *** Error in ./array_random/*repeat/*local/append_tiles/*substitute/max_wh/_minmax_whds/ *** Item substitution '{ mw = w; mh = h; md = d; ms = (...)s#k); ); ([mw,mh,md,ms])[0,2]}': Unrecognized function call 'repeat( l,k, mw = max(mw,w#k); mh = max(mh,h#k); md = max(m(...)' in expression '...repeat( l,k, mw = max(mw,w#k); mh = max(mh,h#k); md = max(m(...)...'.
- Ascii Art : *** Error in ./fx_asciiart/*repeat/*local/(...)/*local/*substitute/max_w/_minmax_whds/ *** Item substitution '{ mw = w; mh = h; md = d; ms = (...)s#k); ); ([mw,mh,md,ms])[0,1]}': Unrecognized function call 'repeat( l,k, mw = max(mw,w#k); mh = max(mh,h#k); md = max(m(...)' in expression '...repeat( l,k, mw = max(mw,w#k); mh = max(mh,h#k); md = max(m(...)...'.
- Dices : *** Error in ./fx_dices/frame_round/*repeat/*local/blend_fade/*substitute/max_whds/_minmax_whds/ *** Item substitution '{ mw = w; mh = h; md = d; ms = (...)s#k); ); ([mw,mh,md,ms])[0,4]}': Unrecognized function call 'repeat( l,k, mw = max(mw,w#k); mh = max(mh,h#k); md = max(m(...)' in expression '...repeat( l,k, mw = max(mw,w#k); mh = max(mh,h#k); md = max(m(...)...'.
- Drawn Montage : *** Error in ./fx_drawn_montage/_fx_drawn_montage/ *** Command 'fill': Unrecognized function call 'repeat( 1,l, I==colors[3*l,3]?( if (x<i(#-2,l,0,0,0), i(#-2(...)' in expression '...repeat( 1,l, I==colors[3*l,3]?( if (x<i(#-2,l,0,0,0), i(#-2(...)...'.
- Grid [Hexagonal] : *** Error in ./fx_imagegrid_hexagonal/*repeat/*local/imagegrid_hexagonal/*repeat/*local/array3d/ *** Command 'arrow3d': Undefined argument '$2', in expression '${1-3}'.
- Make Seamless [Patch-based] : *** Error in ./fx_frame_seamless/frame_seamless/*repeat/(...)/append_tiles/*substitute/max_wh/_minmax_whds/ *** Item substitution '{ mw = w; mh = h; md = d; ms = (...)s#k); ); ([mw,mh,md,ms])[0,2]}': Unrecognized function call 'repeat( l,k, mw = max(mw,w#k); mh = max(mh,h#k); md = max(m(...)' in expression '...repeat( l,k, mw = max(mw,w#k); mh = max(mh,h#k); md = max(m(...)...'.
- etc.

The errors seem most of the time related to the function `repeat`. So I made a quick search and I found some discussions about the error :
- https://www.linux.org/threads/gimp-some-gmic-filters-dont-work.45967/
- https://www.reddit.com/r/krita/comments/p9qft1/help_gmic_in_krita_is_not_working/  giving a link to a video "How to fix Gmic in krita and GIMP easy!" https://www.youtube.com/watch?v=mMnBr-vbkDI (on Windows)
But none of them really helped me.

For information, I was using GIMP installed by snap. I tried other installation methods (using Flatpak and AppImage), but Gimp never found the Plug-in G'Mic, whatever the directory I put it in or whatever the installation method.

Let me know if I can try something else.

Thanks!

Fabien

**Comment by dtschump** [comment]
> I use the last version 3.3.0 of "Ubuntu 22.04 Jammy: Plug-in for GIMP 2.10" with GIMP 2.10.30 on ... Ubuntu 23.04 ... maybe is it the cause of the issue? (But many other filters work anyway)

It seems to me (looking at the screenshot) that you are rather using the quite old version 2.9.6.
This may explain the issue.
Could you try with latest stable version ***3.3.0** ? This issue has been probably fixed since 2.9.6 !

**Comment by fabiengabriel** [comment]
Okay, I got it!

Explanation:

When installing GIMP 2.10.30 via Snap, it includes an old version of G'Mic (the 2.9.6). I found the file "gmic_gimp_qt" in the directory /snap/gimp/393/usr/lib/gimp/2.0/plug-ins/. I did not even check first if there already was an old version of G'Mic before downloading and installing its version 3.3.0. I was just surprised that G'Mic did not appear in the menu when placed in ~/snap/gimp/393/.config/GIMP/2.10/plug-ins, certainly because of a conflict with the old version (so I moved it in another). It was a so obvious clue, but I did not notice!

The issue remains because I do not know how to remove the old plugin version. The system doesnot let me easily remove the file gmic_gimp_qt (even in super user), and I guess this is not a good ideal anyway. But I did not find how to remove it (using snap?).

Do you have an idea?

**Comment by dtschump** [comment]
> Do you have an idea?

Not really no. I'm not using `snap` at all to install things, sorry.

**Comment by fabiengabriel** [comment]
I've finally found a solution:

1. I removed the snap installation of GIMP, purging everything: ``sudo snap remove gimp --purge``
2. I installed GIMP using Flatpak as recommended on https://www.gimp.org/downloads/: ``flatpak install https://flathub.org/repo/appstream/org.gimp.GIMP.flatpakref``
3. I installed G'Mic using Flatpak as proposed here https://www.gimp-forum.net/Thread-GMic-Disappeared-from-GIMP-2-10-24-Flatpak-Installation-Solved: ``flatpak install flathub org.gimp.GIMP.Plugin.GMic`` where I sellected the Branch 2-40.
4. Now I can use the last G'Mic 3.30 version and all the filters seem to work

![Capture d’écran du 2023-09-21 22-13-03](https://github.com/GreycLab/gmic/assets/41618686/1138432f-43e9-4b0d-a48e-1635d2f0f80c)

Thank you anyway for your help.
